### ISSUE TITLE
Element in sequential focus order has visible focus [oj04fd] expectation allows common scenarios to pass even when failing

### ISSUE THREAD

**Issue by giacomo-petri** [auto-wcag/auto-wcag#2368]
I understand that this rule is atomic, and passing it doesn't necessarily mean that rule 2.4.7 is also met. However, there are common scenarios where the focus indicator is invisible, yet the current rule still passes.

A typical example is a top navigation menu or toolbar, where each control uses `outline: none` when focused and triggers a fly-out menu to expand. This expansion changes the pixels on screen, so our rule considers it a pass. However, because `outline: none` removes the visible focus indicator, users can't tell which option is currently focused or which menu item has been expanded.

I also believe this rule currently lacks sufficient examples, so I've added several in https://github.com/act-rules/act-rules.github.io/pull/2369 PR.
Failed Example 3 illustrates the situation I mentioned earlier (even though the description is slightly different). I've marked it as a failure, even though (based on the current wording) it technically passes the rule's expectations.

The PR is currently in draft status, as I'd like to discuss these points with the group before making any changes.

**Comment by dan-tripp-siteimprove** [comment]
You want to fail any case where the changed pixels might not be noticed by the user - am I right?

If I'm right: it seems to me that there are two categories of "changed pixels which might not be noticed by the user": 1) changed pixels which are outside the viewport, and 2) changed pixels that are inside the viewport but are "too far away to be noticed" (which is vague).

But I wonder if we could catch a lot of them this way: 

We change the expectation to "the changed pixels need to be either a) part of the target element (= within its bounding box including CSS margin), or b) on opposite sides of the target element".   And (b) means either: b1) both above and below the target element, or b2) both to the left and to the right of the target element.

Passed Examples 1 and 2 would pass because of (a).

[Passed Example 3](https://www.w3.org/WAI/content-assets/wcag-act-rules/testcases/oj04fd/b44aa294df6f0de62f11f6ae732087d92c90b448.html) would pass because of (b1) and (b2).

[Passed Example 4](https://www.w3.org/WAI/content-assets/wcag-act-rules/testcases/oj04fd/dd9628d86628e285fe99ce98efdacbe441c20ca5.html) would pass because of (b2).

In [your PR 2369](https://github.com/act-rules/act-rules.github.io/pull/2369), Failed Example 3 would still fail and it would be obvious why.

**Comment by Jym77** [comment]
The problem(s) I see with the new Failed Example in #2369 is that small changes would (I think) make it pass WCAG, and I do not really see how to shape the rule to make that difference.
* If the fly-out links had different names, e.g. "ACT Rules link 1", "WCAG link 1", …, then this would pass 2.4.7 as the name in the fly-out links would tell which menu item is focused.
* If the fly-outs were not packed at the left, but left-aligned below their menu item, then again the position would tell which item is focused and pass 2.4.7.

So, while I totally agree that this rule could be stricter, I find it difficult to phrase it in a way that won't create false positives. Thus said, if you have ideas to improve it, definitely go on.

@dan-tripp-siteimprove I'm not sure that "within the bounding box or on opposite sides" is enough too, an arrow on one side of a vertical menu pointing to the focused element, but out of the bounding box of that element would totally work. So, we'd need to get quite a lot of extra complication to not fail those scenarios.

**Comment by carlosapaduarte** [comment]
CG decision: Failed example 3 should be removed because of what we have in the background already ("WCAG does not require that the focus indicator for each [focusable][] element is unique in appearance. Therefore, this rule can pass even if several focus indicators are identical. Such a situation may nonetheless cause confusion and all examples in this rule avoid it.")

All the other examples are okay.

**Comment by giacomo-petri** [comment]
The first sentence of the Understanding document states:

> The purpose of this success criterion is to help a person know which element has the keyboard focus.

When multiple elements appear to share the same focus, users cannot determine which element is actually focused. I understand this isn't part of the normative wording, but it is the very first line of the Understanding, which is meant to clarify the intent.

I've created another, simpler example (not as realistic as the original fly-out scenario, but helpful for illustrating my point): https://codepen.io/giacomo-usablenet/full/QwNrKVy

In this example, the focus indicator is visible, but it's still impossible to tell which element actually has focus.

**Comment by Jym77** [comment]
I don't think anybody is trying to argue that these examples satisfy WCAG 2.4.7. I certainly agree that they are clear failures.

But chiselling the rule to fail them can be difficult. We didn't managed to do it [when we wrote it](https://github.com/act-rules/act-rules.github.io/pull/1329). We have grown and evolved in our rules writing skills since, so maybe we can do better now.

Among others, I feel this rule is already a bit borderline within the format. We had (after writing this) some discussions about [stateful rules](https://github.com/act-rules/act-rules.github.io/issues/1953) which sadly never concluded. I think the current rule (comparing the unfocused and focused state) is already somewhat stateful. If we try and compare several focused states of different focusable elements (to see they are the same and should fail), that will make the rule even more stateful and out of the current format.
(I feel the statefulness of this rule on pages that add focus indicators with `onfocus` listeners adding a class name to the element, Alfa being purely static and not looking at JS doesn't see these indicators).

If you have ideas on how to improve the rule so it can fail these cases, I guess we are all interested ☺️

**Comment by giacomo-petri** [comment]
I think the rule already covers this.
In the last codepen example, there is no visual difference between the state where button 1 is focused (and button 2 is not) and the state where button 2 is focused. If you compare the default, unfocused page to the focused state of button 2, then yes, there is a difference, but that's not only what the applicability and expectation describe:

> For each target element, there is at least one device pixel inside the [scrolling area](https://drafts.csswg.org/cssom-view/#scrolling-area) of the [viewport](https://drafts.csswg.org/css2/#viewport) whose [HSL color value](https://www.w3.org/TR/css-color-3/#hsl-color) is different when the element is [focused](https://www.w3.org/WAI/standards-guidelines/act/rules/oj04fd/#focused) from when it is not.

When button 1 is focused, button 2 is not, and when button 2 becomes focused, there is still no change in its appearance; so IMO button 2 fails the rule.

**Comment by Jym77** [comment]
Good point.

I think it actually raises an ambiguity in the rule… "when [the element] is not [focused]" can means "when nothing is focused", or "when something else is focused", (and possibly "when something is hovered/checked/…"), … and each of these states can have differences between them (as your example shows). The rule is not clear which is the one that should be considered (and we reach both limits of statefulness, and the risk of impracticable "different from all other states" quantifier).

Pushing it to the limit, an element that has the same `:focus` and `:hover` rendering (but different from the default state) could be said as not having any pixel difference between its focused and (one of its, namely the hovered one) non-focused states. But I'm not sure it should fail the SC (I'm not sure it should pass it either 😅).
